dart/tool/dart/generate_dart.dart

```diff
@@ -453,7 +483,108 @@ typedef ServiceCallback = Future<Map<String, dynamic>> Function(
     });
     gen.writeln('};');
     gen.writeln();
-    gen.writeStatement('class VmService {');
+
+    // The service interface, both servers and clients implement this.
+    gen.writeStatement('''
+abstract class VmServiceInterface {
+  /// Returns the stream for a given stream id.
+  ///
+  /// This is not a part of the spec, but is needed for both the client and
+  /// server to get access to the real event streams.
+  Stream<Event> onEvent(String streamId);
+''');
+    methods.forEach((m) {
+      m.generateDefinition(gen);
+      gen.write(';');
+    });
+    gen.write('}');
+    gen.writeln();
+
+    // The server class, takes a VmServiceInterface and delegates to it
+    // automatically.
+    gen.write('''
+  class VmServer {
+    final Stream<Map<String, Object>> requestStream;
+    final StreamSink<Map<String, Object>> responseSink;
+    final VmServiceInterface serviceImpl;
+
+    /// Manages streams for `streamListen` and `streamCancel` requests.
+    final _streamSubscriptions = <String, StreamSubscription>{};
+
+    VmServer(this.requestStream, this.responseSink, this.serviceImpl) {
+      requestStream.listen(_delegateRequest);
+    }
+
+    void _delegateRequest(Map<String, Object> request) async {
+      try {
+        var method = request['method'];
+        if (method == null) {
+          throw RPCError(null, -32600, 'Invalid Request', request);
+        }
+        var params = request['params'] as Map;
+        Response response;
+        switch(method) {
+    ''');
+    methods.where((m) => !m.isUndocumented).forEach((m) {
+      gen.writeln("case '${m.name}':");
+      if (m.name == 'streamListen') {
+        gen.writeln(_streamListenCaseImpl);
+      } else if (m.name == 'streamCancel') {
+        gen.writeln(_streamCancelCaseImpl);
+      } else {
+        gen.write("response = await serviceImpl.${m.name}(");
+        // Positional args
+        m.args.where((arg) => !arg.optional).forEach((arg) {
+          gen.write("params['${arg.name}'], ");
+        });
+        // Optional named args
+        var namedArgs = m.args.where((arg) => arg.optional);
+        if (namedArgs.isNotEmpty) {
+          namedArgs.forEach((arg) {
+            gen.writeln("${arg.name}: params['${arg.name}'], ");
+          });
+        }
+        gen.writeln(");");
+      }
+      gen.writeln('break;');
+    });
+    // Throw for unrecognized methods.
+    gen.writeln('default:');
```

Comment:
This would be for cases where the client is trying to use a newer version of the protocol than the generated code (and server impl) supported?

Reply:
Generally yes - although it could also happen if you attempt to use a private method which this doesn't support, or just make a typo or something in your method name.

Note that this exception will get converted into a proper jsonrpc2 error response.